LFS-25: As a bioinformatician, I can upload somatic variant files as CSVs in bulk

veronikaslc opened the pull request:
LFS-449: New questionnaire: SomaticVariants
LFS-450: Page for bulk variants upload
LFS-451: New answer type: SomaticVariantsAnswer
LFS-452: New answer component: SomaticVariantsQuestion

Commit:
LFS-25: As a bioinformatician, I can upload somatic variant files as CSVs in bulk
Work in progress

sdumitriu committed:
LFS-25: As a bioinformatician, I can upload somatic variant files as CSVs in bulk
Added the variants module to the distribution package

sdumitriu committed:
LFS-25: As a bioinformatician, I can upload somatic variant files as CSVs in bulk
Fixed the nodetype definition

Commit:
LFS-25: As a bioinformatician, I can upload somatic variant files as CSVs in bulk
Work in progress

Commit:
LFS-449: New questionnaire: SomaticVariants

Commit:
LFS-450: Page for bulk variants upload

veronikaslc committed:
LFS-452: New answer component: SomaticVariantsQuestion
Work in ptogress

sashaandjic commented:
When I try to associate tumour region with tumour, I get the blank screen.
Also, if there is no tumour to choose, I should be able to create a new one. Perhaps this functionality is in PR for LFS-426/LFS-458
https://github.com/ccmbioinfo/lfs/pull/226#pullrequestreview-437578940 ?

sashaandjic commented:
Is there any test file that I can use to test upload?

veronikaslc commented:
@sashaandjic  >> When I try to associate tumour region with tumour, I get a blank screen.
I am not sure which place are you talking about. In my PR there was just a file upload and a SomaticVariants questionnaire with static display of uploaded file.

veronikaslc commented:
@sashaandjic I used arbitrary files to test, like those you can download from here https://support.spatialkey.com/spatialkey-sample-csv-data/

sashaandjic commented:
> @sashaandjic >> When I try to associate tumour region with tumour, I get a blank screen.
> I am not sure which place are you talking about. In my PR there was just a file upload and a SomaticVariants questionnaire with static display of uploaded file.

I was trying to create a new Somantic Variants Data form. As soon as I chose the subject the form goes white. The same happens when I click on any existing Somantic Variants Data forms.
![image](https://user-images.githubusercontent.com/23243244/85903829-4b738380-b7d5-11ea-8875-ebc2d41c5f5d.png)

veronikaslc commented:
As per [LFS-452](https://phenotips.atlassian.net/browse/LFS-452) for a start  this form is only auto-created when file is uploaded. The actual UI will be detailed later, we do not create the form itself yet as far as I understood. Main focus of the PR was LFS-450, the bulk file upload.

marta- commented on modules/variants/src/main/frontend/src/variantsContainer.jsx:
```diff
@@ -0,0 +1,700 @@
+    (function loop(i) {
+        if (i < chosenFiles.length) new Promise((resolve, reject) => {
+          let file = chosenFiles[i];
+          if (!(/^(.+)_(.+)_(.+).csv$/.test(file.name))) {
```
Tumor region should be optional

marta- commented on modules/variants/src/main/frontend/src/variantsContainer.jsx:
```diff
@@ -0,0 +1,700 @@
+          file.tumor       = {}
+          file.tumor.id    = parsed[1];
+          file.region      = {}
+          file.region.id   = parsed[2];
```
I would redesign this code to leverage the subject type hierarchy. I would make variants a subject type that can gave Tumor or Tumor region as parents, and ensure the code relies on this configured hierarchy rather than hardcoding the related subjects.

veronikaslc replied:
Can you please give a short example of such structure

marta- commented on modules/variants/src/main/frontend/src/variantsContainer.jsx:
```diff
@@ -0,0 +1,700 @@
+  };
+
+  // Change of subject id implies reset patient subject info and re-fetching all data
+  let setSubject = (id, fileName) => {
```
There's a lot of duplicated code in the next ~50 lines.

marta- commented on modules/variants/src/main/frontend/src/variantsContainer.jsx:
```diff
@@ -0,0 +1,700 @@
+      ) }
+      { error && <Typography color='error'>{error}</Typography> }
+      <label htmlFor="contained-button-file">
+        <IconButton color="primary" aria-label="upload picture" component="span">
```
I did not realize this was a button when I first saw it. My attention was fully captured by thhe non-functional Upload button.

 I suggest:
- making the file selection aria stand off more
- displaying the Upload button only when there's something to upload.

This is an example of a file upload ui with a more visible upload area.
![Screenshot_20200707-103054](https://user-images.githubusercontent.com/651980/86797201-62b83980-c03d-11ea-8024-cd985197ea17.jpg)

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/SomaticVariantsQuestion.jsx:
```diff
@@ -0,0 +1,58 @@
+//
```
Can this file be part of the `variants` module instead?

veronikaslc replied:
Why? `VocabularyQuestion.jsx` is not part of `vocabularies`.

marta- replied:
I agree it should be in variants (and VocabularyQuestion should be in vocabularies). The module should contain all the components it needs, and the ui components should not be scattered in other ui modules. But I'm fine with leaving it here for now and reorganize all the questions later.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/SomaticVariantsQuestion.jsx:
```diff
@@ -0,0 +1,58 @@
+
+function SomaticVariantsQuestion(props) {
+  let { existingAnswer, classes, ...rest } = props;
+  let file = existingAnswer[1][Object.keys(existingAnswer[1]).find(key => existingAnswer[1][key]["jcr:primaryType"] === "nt:file")];
```
```suggestion
  let file = existingAnswer?.[1][Object.keys(existingAnswer[1]).find(key => existingAnswer[1][key]["jcr:primaryType"] === "nt:file")];
```

sdumitriu commented on modules/variants/src/main/frontend/package.json:
```diff
@@ -0,0 +1,54 @@
+      "webpack-assets-manifest": "3.1.1",
+      "webpack-cli": "^3.3.2"
+    },
+    "dependencies": {
```
I wonder how many of these are actually needed by the code in this module.

sdumitriu commented on modules/variants/src/main/frontend/package.json:
```diff
@@ -0,0 +1,54 @@
+    "repository": {
+      "type": "git",
+      "url": "https://github.com/ccmbioinfo/lfs.git",
+      "directory": "modules/homepage/src/main/frontend"
```
```suggestion
      "directory": "modules/variants/src/main/frontend"
```

sdumitriu commented on modules/variants/src/main/resources/SLING-INF/nodetypes/variants.cnd:
```diff
@@ -0,0 +1,38 @@
+[lfs:SomaticVariantsAnswer] > lfs:Answer
+
+  // Hardcode the resource type.
+  - sling:resourceType (STRING) = "lfs/SomaticVariantAnswer" mandatory autocreated protected
```
```suggestion
  - sling:resourceType (STRING) = "lfs/SomaticVariantsAnswer" mandatory autocreated protected
```

sdumitriu commented on questionnaires/clinical-data/src/main/resources/SLING-INF/content/SomaticVariants.json:
```diff
@@ -0,0 +1,12 @@
+{
```
This file exists twice, you should delete this one and keep the one in the `questionnaires/somatic-variants` module.

veronikaslc replied:
@sdumitriu The other file does not have line 4 in it. Should I copy this line there? `"jcr:reference:requiredSubjectTypes": ["/SubjectTypes/TumorRegion"],`

sdumitriu replied:
Yes.

sdumitriu commented on questionnaires/somatic-variants/src/main/resources/SLING-INF/content/SomaticVariants.json:
```diff
@@ -0,0 +1,10 @@
+  "title": "Somatic Variants Data",
+  "file": {
+    "jcr:primaryType": "lfs:Question",
+    "text": "Somatic Variant File",
```
```suggestion
    "text": "Somatic Variants File",
```

sdumitriu commented on questionnaires/somatic-variants/src/main/resources/SLING-INF/content/SomaticVariants.json:
```diff
@@ -0,0 +1,10 @@
+    "jcr:primaryType": "lfs:Question",
+    "text": "Somatic Variant File",
+    "dataType": "somaticVariantsFile",
+    "maxAnswers": 1
```
```suggestion
    "minAnswers": 1,
    "maxAnswers": 1
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -231,7 +231,11 @@ const questionnaireStyle = theme => ({
         disable: "flex"
-    }
+    },
+    fileInfo: {
```
Delete this, it is already in the new module.

veronikaslc replied:
It is needed to display the file info on the form.

Commit:
LFS-450: Page for bulk variants upload
Added a drop zone for file upload, relaxed mandatory tumor region restriction for filename

veronikaslc commented:
@marta- Added Drag&drop area

veronikaslc committed:
LFS-452: New answer component: SomaticVariantsQuestion

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

veronikaslc committed:
LFS-25: As a bioinformatician, I can upload somatic variant files as CSVs in bulk

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

veronikaslc committed:
LFS-25: As a bioinformatician, I can upload somatic variant files as CSVs in bulk

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

veronikaslc committed:
LFS-449: New questionnaire: SomaticVariants

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

veronikaslc committed:
LFS-449: New questionnaire: SomaticVariants

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

sashaandjic commented:
Small complaint: "tumor" should be spelled as "tumour" ( I make this mistake all the time too)
![image](https://user-images.githubusercontent.com/23243244/87496280-40b44d80-c621-11ea-9ec8-4c50339e4051.png)

Also, the error does not seem to clear up once properly named file is uploaded.

Commit:
LFS-25: As a bioinformatician, I can upload somatic variant files as CSVs in bulk
Addressed codereview comments

sashaandjic commented:
After successful upload of file called `qwerty_asdfg tt.csv`, the subject `qwerty` gets created.
![image](https://user-images.githubusercontent.com/23243244/87514573-715bad80-c648-11ea-9b7f-b0fbd3b7fcdc.png)

However subject is not displayed when I open the form.
![image](https://user-images.githubusercontent.com/23243244/87514189-d5ca3d00-c647-11ea-96b9-0d2d8b50bf77.png)

Commit:
LFS-25: As a bioinformatician, I can upload somatic variant files as CSVs in bulk
Addressed codereview comments

sdumitriu committed:
LFS-25: As a bioinformatician, I can upload somatic variant files as CSVs in bulk
Fixed bug: subjects don't properly display their hierarchy

sdumitriu committed:
Revert "LFS-25: As a bioinformatician, I can upload somatic variant files as CSVs in bulk"

This reverts commit cb1be31f5af0b520d5fc95c7dc617bc87ecaba2e.

sdumitriu committed:
LFS-25 / LFS-449: New questionnaire: SomaticVariants
Keep only one copy of the questionnaire

sashaandjic commented:
I uploaded file phenomecentral_telethon_measurements_2020-07-22_17-15.csv and got the follwing:
- **pheneomecentral**- subject type of patient
- **telethon** - subject type of tumor associated with **phenomecentral**
- **measurements** - subject type of tumor region associated with **telethon**
![image](https://user-images.githubusercontent.com/23243244/91511337-72ba0000-e8ad-11ea-9310-f5f1080915c1.png)
- somatic variants data form associated with newly created tumor region
![image](https://user-images.githubusercontent.com/23243244/91511389-8ebda180-e8ad-11ea-8a6a-a91cac836121.png)

sashaandjic commented:
Problem: importing file a_bcd.csv created blank tumor region that I cannot access:
![image](https://user-images.githubusercontent.com/23243244/91524300-dd2e6880-e8cc-11ea-9e70-e5f6cfc75574.png)
I was expecting an error stating :**File name test.csv does not follow the name convention <Subject>_<tumour nb>***.csv**

sashaandjic commented:
@sdumitriu  Question: Importing identical file twice, results in a new Somatic variant data form. Is this correct?

sashaandjic commented:
Sometimes I can select and upload multiple files with success. However sometimes I can't even select the multiple files.
![image](https://user-images.githubusercontent.com/23243244/91589964-940c0200-e928-11ea-81a1-529c810a4ad1.png)

@sdumitriu reported problems with concurrency when uploading multiple files. It is logged in a separate task https://phenotips.atlassian.net/browse/LFS-535?filter=16096

sdumitriu commented:
Yes.

On Fri, Aug 28, 2020, 11:53 Sasha <notifications@github.com> wrote:

> @sdumitriu <https://github.com/sdumitriu> Question: Importing identical
> file twice, results in a new Somatic variant data form. Is this correct?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ccmbioinfo/lfs/pull/240#issuecomment-682759711>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAVUV3VGVY2AXASU5YNPRDSC7HIZANCNFSM4OJBVEPQ>
> .
>

marta- commented on modules/variants/src/main/frontend/src/variantsContainer.jsx:
```diff
@@ -0,0 +1,724 @@
+  }
+}));
+
+export default function variantsContainer() {
```
Component names should start with an uppercase letter. Additionally, this component is about variant _files_, not variants, and should be renamed accordingly to avoid confusions.

marta- commented on modules/variants/src/main/frontend/src/variantsContainer.jsx:
```diff
@@ -0,0 +1,724 @@
+          file.subject     = {}
+          file.subject.id  = parsed[0];
+          file.tumor       = {}
+          file.tumor.id    = parsed[1];
```
It is possible to get an empty tumor id, e.g. with a file named "patientid__02.csv". The code should protect against this.

marta- commented on modules/variants/src/main/frontend/src/dragAndDrop.jsx:
```diff
@@ -0,0 +1,120 @@
+import { IconButton, Typography } from "@material-ui/core";
+import AttachFile from '@material-ui/icons/AttachFile';
+
+class DragAndDrop extends React.Component {
```
All newly added components should be functional components, in order to have a consistent code style.

marta- commented on modules/variants/src/main/frontend/src/dragAndDrop.jsx:
```diff
@@ -0,0 +1,120 @@
+		    type="file"
+		    accept=".csv"
+		    name="*"
+            multiple
```
The components should have a property that specifies whether multiple files are accepted, and this argument should be set according to that property.

marta- commented on modules/variants/src/main/frontend/src/variantsContainer.jsx:
```diff
@@ -0,0 +1,724 @@
+    margin: "0",
+    borderRadius: "2px",
+  },
+  active: {
```
There's a surprising volume of custom styling from here on, including hardcoded sizes and colors that don't rely on the theme spacing. We're aiming for minimalist styling and using the material ui components and theme as much as possible (without really sacrificing the look). Is there a way the style can be rewritten to rely more on the theme?

marta- commented on modules/variants/src/main/frontend/src/variantsContainer.jsx:
```diff
@@ -0,0 +1,724 @@
+        </Grid>
+      ) }
+
+      <DragAndDrop
```
There is a bug with uploading multiple files at the moment, the responses are not handled in the correct order. In this first iteration, add a property to the `DragAndDrop` component allowing to configure whether the component allows multiple upload, and set it to false here to avoid bugs. Please do not delete the code handling the multiple uploads yet, as it will likely not require major changes in order to work correctly. Only restrict to uploading one file at the time.

Commit:
LFS-25 / LFS-449: New questionnaire: SomaticVariants
Addressed codereview comments

sashaandjic commented:
Should the required name convention be _<Subject>_<Tumour nb>***.csv_ or _<Subject>_<Tumor>_***.csv? (underscore after tumor)
I am stil able to import files like 1_2.csv, which result in blank tumor region subject being created.

Everything else is working as expected.

veronikaslc commented:
@sashaandjic I'm not sure too. By [LFS-449 specs](https://phenotips.atlassian.net/browse/LFS-449) TumorRegion subject type is required. By the codereview comments - not required.

Commit:
LFS-25 / LFS-449: New questionnaire: SomaticVariants
Addressed codereview comments

marta- commented on modules/variants/src/main/frontend/src/variantsContainer.jsx:
```diff
@@ -0,0 +1,751 @@
+  }
+}));
+
+export default function VariantsContainer() {
```
As I specified in my previous comment, please rename this to clarify that it's a container for variant _files_, not for variants, e.g. `VariantFilesContainer`.

marta- reviewed (changes requested):
Looks fine, I just have one comment about renaming a component, after which I can approve it.

sashaandjic reviewed (approved):
tested

Commit:
LFS-25 / LFS-449: New questionnaire: SomaticVariants
Changed of VariantsContainer to VariantFilesContainer

sashaandjic commented:
@marta- If I upload file named abc_123.csv, I get 2 subject created: patient abc and tumor 123. 
![image](https://user-images.githubusercontent.com/23243244/93115611-16672680-f68a-11ea-8df0-0f9b866747ed.png)
Since there is no tumor region the Somatic Variants Data file gets associated with the tumor. Is this correct?
![image](https://user-images.githubusercontent.com/23243244/93115667-2b43ba00-f68a-11ea-895d-6e38d806c564.png)

marta- commented:
> Since there is no tumor region the Somatic Variants Data file gets associated with the tumor. Is this correct?

Yep, it's perfect.
